carla/evaluation/utils.py

```diff
@@ -0,0 +1,136 @@
+        return violations
+    else:
+        if doubled_elements is None:
+            return violations
+        else:
+            return violations - len(accounting_factor)
+
+
+def success_rate_and_indices(counterfactuals: pd.DataFrame):
+    """
+    Used to indicate which counterfactuals should be dropped (due to lack of success indicated by NaN).
+    Also computes percent of successfully found counterfactuals
+    :param counterfactuals: pd df, where NaNs indicate 'no counterfactual found' [df should contain no object values)
+    :return: success_rate, indices
+    """
+
+    # Success rate & drop not successful counterfactuals & process remainder
+    success_rate = (counterfactuals.dropna().shape[0]) / counterfactuals.shape[0]
+    counterfactual_indices = np.where(
+        np.logical_not(np.any(np.isnan(counterfactuals.values), axis=1))
+    )[0]
+
+    return success_rate, counterfactual_indices
+
+
+def redundancy(factual: np.array, counterfactual: np.array, ml_model: MLModelCatalog):
+    """
+    Redundency metric which looks for 'none-essential' changes
+    :param factual: np.array with original instance
+    :param counterfactual: np.array with counterfactual instance
+    :param model: pytorch model
+    :return: scalar, number of unnecessary changes
+    """
+    red = 0
+
+    # get model prediction and cast it from tensor to float
+    if ml_model.backend == "pytorch":
+        pred_f = round(
+            ml_model.predict(torch.from_numpy(factual).float())
+            .detach()
+            .numpy()
+            .squeeze()
+            .reshape(1)[0]
+        )
+        pred_cf = round(
+            ml_model.predict(torch.from_numpy(counterfactual).float())
+            .detach()
+            .numpy()
+            .squeeze()
+            .reshape(1)[0]
+        )
+    elif ml_model.backend == "tensorflow":
+        pred_f = ml_model.predict(factual)
+        pred_f = np.argmax(pred_f, axis=1)
+        pred_cf = ml_model.predict(counterfactual)
+        pred_cf = np.argmax(pred_cf, axis=1)
+    else:
+        raise NotImplementedError()
+
+    if pred_f != pred_cf:
+        for i in range(factual.shape[1]):
+            if factual[0][i] != counterfactual[0][i]:
+                temp_cf = np.copy(counterfactual)
+
+                # reverse change in counterfactual and predict new label
+                temp_cf[0][i] = factual[0][i]
+                if ml_model.backend == "pytorch":
+                    pred_temp_cf = round(
+                        ml_model.predict(torch.from_numpy(temp_cf).float())
+                        .detach()
+                        .numpy()
+                        .squeeze()
+                        .reshape(1)[0]
+                    )
+                elif ml_model.backend == "tensorflow":
+                    pred_temp_cf = ml_model.predict(temp_cf)
+                    pred_temp_cf = np.argmax(pred_temp_cf, axis=1)
+                else:
+                    raise NotImplementedError()
```

Comment:
Same as above, we could use `predict_proba `and `argmax `to do this independent of the backend.

Reply:
Changed in refactoring commit.